src/lib.rs

```diff
@@ -0,0 +1,989 @@
+            let min: T = self.data.iter().fold(T::zero(), |acc, x| if acc < *x { acc } else { *x });
+            return Tensor::new(&Shape::new(vec![1]), &[min]);
+        }
+
+        // Create new tensor with right shape
+        let new_shape = Shape::new(remaining_dims);
+        let max: T = self.data.iter().fold(T::zero(), |acc, x| if acc > *x { acc } else { *x });
+        let mut t: Tensor<T> = Tensor::fill(&new_shape, max);
+
+        for target in IndexIterator::new(&new_shape.dims) {
+            let min_iter = IndexIterator::new(&removing_dims);
+            for min_index in min_iter {
+                let mut indices = target.clone();
+                for (i, &axis) in axes.iter().enumerate() {
+                    indices.insert(axis, min_index[i]);
+                }
+
+                if self.get_element(&indices) < t.get_element(&target) {
+                    t.set_element(&target, *self.get_element(&indices));
+                }
+            }
+        }
+
+        t
+    }
+
+    // Tensor Product
+    // Consistent with numpy.tensordot(a, b, axis=0)
+    pub fn prod(&self, other: &Tensor<T>) -> Tensor<T> {
+        let mut new_dims = self.shape.dims.clone();
+        new_dims.extend(&other.shape.dims);
+        let new_shape = Shape::new(new_dims);
+
+        let mut new_data = Vec::with_capacity(self.size() * other.size());
+        for &a in &self.data {
+            for &b in &other.data {
+                new_data.push(a * b);
+            }
+        }
+
+        Tensor::new(&new_shape, &new_data)
+    }
+
+    /// For the maths see: https://bit.ly/3KQjPa3
+    fn calculate_index(&self, indices: &[usize]) -> usize {
+        let mut index = 0;
+        for k in 0..self.shape.len() {
+            let stride = self.shape.dims[k+1..].iter().product::<usize>();
+            index += indices[k] * stride;
+        }
+        index
+    }
+
+    fn assert_indices(&self, indices: &[usize]) -> Result<(), String> {
+        if indices.len() != self.shape.len() {
+            return Err(format!("Incorrect number of dimensions ({} vs {}).", indices.len(), self.shape.len()));
+        }
+        for (i, &index) in indices.iter().enumerate() {
+            if index >= self.shape.dims[i] {
+                return Err(format!("Index out of bounds for dimension {}", i));
+            }
+        }
+        Ok(())
+    }
+
+}
+
+// Element-wise Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<T> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: T) -> Tensor<T> {
+        let mut result = Tensor::zeros(&self.shape);
+        for i in 0..self.size() {
+            result.data[i] = self.data[i].clone() * rhs;
+        }
+        result
+    }
+}
+
+// Vector/Matrix Multiplication
+impl<T: Num + PartialOrd + Copy> Mul<Tensor<T>> for Tensor<T> {
+    type Output = Tensor<T>;
+
+    fn mul(self, rhs: Tensor<T>) -> Tensor<T> {
+    if self.shape.len() == 1 && rhs.shape.len() == 1 {
```

Comment:
I think a lot these cases could just be merged? At least, I think that's the case if the code differentiates between row and column vectors.

Reply:
Yep, definitely needs a rewrite.

Reply:
They've been split across the different methods in `DynamicVector`, `DynamicMatrix` and `DynamicTensor`.